File: server/src/main/java/com/grpc/empuje_comunitario/infrastructure/grpc/UserGrpcService.java
package com.grpc.empuje_comunitario.infrastructure.grpc;

import com.grpc.empuje_comunitario.application.user.UserRepositoryImpl;
import com.grpc.empuje_comunitario.domain.user.Role;
import com.grpc.empuje_comunitario.domain.user.User;
import com.grpc.empuje_comunitario.proto.CreateUserRequest;
import com.grpc.empuje_comunitario.proto.CreateUserResponse;
import com.grpc.empuje_comunitario.proto.UserServiceGrpc;
import io.grpc.Status;
import io.grpc.stub.StreamObserver;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.grpc.server.service.GrpcService;
import org.springframework.transaction.annotation.Transactional;


@GrpcService
public class UserGrpcService extends UserServiceGrpc.UserServiceImplBase {

    private final UserRepositoryImpl userRepository;

    @Autowired
    public UserGrpcService(UserRepositoryImpl userRepository) {
        this.userRepository = userRepository;
    }

    @Override
    @Transactional
    public void createUser(CreateUserRequest request, StreamObserver<CreateUserResponse> responseObserver) {
        try {
            User user =  User.create(
                    request.getUsername(),
                    request.getName(),
                    request.getLastname(),
                    request.getPhone(),
                    request.getEmail(),
                    Role.valueOf(request.getRole())
            );
            userRepository.create(user);

            CreateUserResponse response = CreateUserResponse.newBuilder()
                    .setSuccess(true)
                    .setMessage("User created successfully")
                    .build();

            responseObserver.onNext(response);
            responseObserver.onCompleted();

        } catch (IllegalArgumentException e) {
            System.err.println("Validation error: " + e.getMessage());
            Status status = Status.INVALID_ARGUMENT.withDescription(e.getMessage());
            responseObserver.onError(status.asRuntimeException());
        } catch (Exception e) {
            System.err.println("Unexpected error: " + e.getMessage());
            Status status = Status.INTERNAL.withDescription("Error: " + e.getMessage());
            responseObserver.onError(status.asRuntimeException());
        }
    }
}

File: server/src/main/java/com/grpc/empuje_comunitario/infrastructure/grpc/UserGrpcService.kt
package com.grpc.empuje_comunitario.infrastructure.grpc

import com.grpc.empuje_comunitario.application.user.UserRepositoryImpl
import com.grpc.empuje_comunitario.domain.user.Role
import com.grpc.empuje_comunitario.domain.user.User
import com.grpc.empuje_comunitario.proto.CreateUserRequest
import com.grpc.empuje_comunitario.proto.CreateUserResponse
import com.grpc.empuje_comunitario.proto.UserServiceGrpc
import io.grpc.Status
import io.grpc.stub.StreamObserver
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.grpc.server.service.GrpcService
import org.springframework.transaction.annotation.Transactional

@GrpcService
open class UserGrpcService @Autowired constructor(
    private val userRepository: UserRepositoryImpl
) : UserServiceGrpc.UserServiceImplBase() {

    @Transactional
    override fun createUser(
        request: CreateUserRequest,
        responseObserver: StreamObserver<CreateUserResponse>
    ) {
        try {
            val user = User.create(
                request.username,
                request.name,
                request.lastname,
                request.phone,
                request.email,
                Role.valueOf(request.role)
            )
            val result = userRepository.create(user)

            val response = CreateUserResponse.newBuilder()
                .setSuccess(result == com.grpc.empuje_comunitario.domain.Result.SUCCESS)
                .setMessage("User created successfully")
                .build()

            responseObserver.onNext(response)
            responseObserver.onCompleted()
        } catch (e: IllegalArgumentException) {
            System.err.println("Validation error: ${e.message}")
            val status = Status.INVALID_ARGUMENT.withDescription(e.message)
            responseObserver.onError(status.asRuntimeException())
        } catch (e: Exception) {
            System.err.println("Unexpected error: ${e.message}")
            val status = Status.INTERNAL.withDescription("Error: ${e.message}")
            responseObserver.onError(status.asRuntimeException())
        }
    }
}